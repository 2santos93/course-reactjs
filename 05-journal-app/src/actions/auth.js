import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from "firebase/auth";
import Swal from 'sweetalert2'

import { googleAuthProvider } from "../Firebase/firebaseConfig";

import { types } from "./../Types/types";
import { desactiveNote, setNotes } from "./note";
import { finishLoading, startLoading } from './ui';

const auth = getAuth();
// functions that return functions are called currying (the action as such)
export const beforeLogin = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    signInWithEmailAndPassword(auth, email, password)
      .then(({user}) => {
        dispatch(login(user.uid, user.displayName));
      })
      .catch((e) => {
        Swal.fire('Fail', e.message, 'error');
      })
      .finally(() => {
        dispatch(finishLoading());
      });
  };
};

export const beforeRegister = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password).then(
      async ({ user }) => {
        await updateProfile(user, { displayName: name }); // for udpate the name
        dispatch(login(user.uid, user.displayName));
      }
    );
  };
};

export const beforeGoogleLogin = () => {
  return (dispatch) => {

    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      })
      .catch((e) => {
        Swal.fire('Fail', e.message, 'error');
      });
  };
};


export const beforeLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
    dispatch(setNotes([]));
    dispatch(desactiveNote());
  }
}

export const login = (uid, name) => {
  return {
    type: types.login,
    payload: {
      uid,
      name,
    },
  };
};

export const logout = () => {
  return {
    type: types.logout
  }
}
