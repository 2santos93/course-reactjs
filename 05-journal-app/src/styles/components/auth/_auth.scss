.auth__main {
  align-items: center;
  background-color: $primary;
  display: flex;
  justify-content: center;
  margin: 0px;
  height: 100vh;
  width: 100vw;
}

.auth__box-main {
  background-color: white;
  width: 250px;
  box-shadow: 0 0 15px $dark-grey;
  border-radius: 5px;
  padding: 20px;
}



