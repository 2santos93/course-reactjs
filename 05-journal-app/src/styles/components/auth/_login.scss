.auth__header-login {
  color: darken($primary, 15%)
}

.auth__input {
  border: none;
  border-bottom: 1px solid $light-grey;
  margin-top: 15px;
  width: 100%;
  font-size: 16px;
  color: $dark-grey;

  &:focus{
    outline: none;
    border-bottom: 1px solid $primary;
  }
}

.auth__button{
  border-radius: 4px;
  height: 36px;
  width: 100%;
  margin-top: 20px;
  background-color: lighten($primary, 5%);
  color: $white;
  cursor: pointer;
  border: 1px solid darken($light-grey, 5%);
  box-shadow: 0 0px 2px rgba(0, 0, 0, 0.1);

  &:hover{
    background-color: $primary;
  }

  &:disabled{
    background-color: lighten($primary, 15%);
    cursor: default;
  }
}

.auth__footer {
  margin-top: 20px;

  a {
    text-decoration: none;

    &:active {
      color: black;
      border: 1px solid darken($light-grey, 5%);
    }
  }
}