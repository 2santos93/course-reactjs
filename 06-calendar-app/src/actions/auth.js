import { request } from '../helpers/requets';
import { methods } from '../types/methods';
import { types } from '../types/types';

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const loginData = await request(
        `${process.env.REACT_APP_API_URL}/auth/login`,
        methods.POST,
        { email, password }
      );

      const { uid, name, token } = loginData;

      localStorage.setItem('token', token);
      localStorage.setItem('token-time', new Date().getTime());
      dispatch(login({ uid, name }));
    } catch (e) {
      alert(`login error ${e.message}`);
    }
  };
};

export const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    try {
      const response = await request(
        `${process.env.REACT_APP_API_URL}/auth/register`,
        methods.POST,
        { email, password, name }
      );

      // if (!response.ok) throw new Error(response.error);

      const { uid, name: userName, token } = response;

      localStorage.setItem('token', token);
      localStorage.setItem('token-time', new Date().getTime());

      dispatch(register({ uid, name: userName }));
    } catch (e) {
      alert(`Register error ${e.message}`);
    }
  };
};

export const register = (user) => ({
  type: types.authRegister,
  payload: user,
});

export const checkingStart = () => {
  return async (dispatch) => {
    try {
      const user = await request(
        `${process.env.REACT_APP_API_URL}/auth/renew`,
        methods.GET,
        null,
        true
      );

      if (!user.uid) throw new Error('Not token found');

      const { uid, name, token } = user;

      localStorage.setItem('token', token);
      localStorage.setItem('token-time', new Date().getTime());
      dispatch(login({ uid, name }));
    } catch (e) {
      dispatch(checking());
    }
  };
};

export const checking = () => {
  return {
    type: types.authChecking,
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.removeItem('token');
    localStorage.removeItem('token-time');

    dispatch(logout());
  };
};

export const logout = () => ({
  type: types.authLogout,
});
