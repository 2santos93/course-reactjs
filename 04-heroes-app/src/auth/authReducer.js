import { AUTH } from "../types/Auth";

export const authReducer = (state = {}, action) => {
  switch(action.type){
    case AUTH.LOGIN:
      return {
        ...action.payload,
        logged: true
      }
    case AUTH.LOGOUT:
      return {
        logged: false,
      }
    default:
      return state;
  }
}