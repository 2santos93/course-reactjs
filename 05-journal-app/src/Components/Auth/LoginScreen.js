import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { beforeGoogleLogin, beforeLogin } from "../../actions/auth";
import { useForm } from "./../../hooks/useForm";

export const LoginScreen = () => {
  const [formvalues, handleInputChange] = useForm();
  const {loading} = useSelector(state => state.ui)
  const dispatch = useDispatch();
  const { email, password } = formvalues;

  const onSubmitHandler = (e) => {
    e.preventDefault();
    dispatch(beforeLogin(email, password));
  };

  const onSubmitHandlerGoogle = () => {
    dispatch(beforeGoogleLogin());
  };

  return (
    <div>
      <h1 className="auth__header-login">Login</h1>
      <form onSubmit={onSubmitHandler} >
        <input
          type="email"
          name="email"
          className="auth__input"
          placeholder="Email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          className="auth__input"
          placeholder="Password"
          value={password}
          name="password"
          onChange={handleInputChange}
        />
        <button type="submit" className="auth__button" disabled={loading} >
          Login
        </button>
      </form>

      <div className="google-btn" onClick={onSubmitHandlerGoogle}>
        <div className="google-icon-wrapper">
          <img
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
            alt="google button"
          />
        </div>
        <p className="btn-text">
          <b>Sign in with google</b>
        </p>
      </div>

      <div className="auth__footer">
        <Link to="/auth/register">Create new account</Link>
      </div>
    </div>
  );
};
