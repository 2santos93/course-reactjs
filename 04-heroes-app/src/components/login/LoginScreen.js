import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext'
import { AUTH } from '../../types/Auth';

export const LoginScreen = ({history}) => {

  const {dispatch} = useContext(AuthContext);

  const loginHandler = () => {
    // history.push('/marvel');
    const loginAction = {
      type: AUTH.LOGIN,
      payload: {
        name: 'nelson'
      }
    }
    
    dispatch(loginAction);
    
    history.replace('/');
  }

  return (
    <>
      <h1>Login</h1>
      <hr />

      <button
        className='btn btn-primary'
        onClick={loginHandler}
      >
        Login
      </button>
    </>
  )
}
