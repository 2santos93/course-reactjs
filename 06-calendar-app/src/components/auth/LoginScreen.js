import React from 'react';
import './login.css';
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';

export const LoginScreen = () => {
  const dispatch = useDispatch();

  const onLoginSubmitHandler = (e) => {
    e.preventDefault();
    const email = document.getElementById('lemail');
    const password = document.getElementById('lpassword');
    const emailValue = email.value;
    const passwordValue = password.value;

    if (emailValue.trim().length < 2 || passwordValue.trim().length < 6)
      return alert('error in login form');

    dispatch(startLogin(emailValue, passwordValue));
    email.value = '';
    password.value = '';
  };

  const onSubmitRegisterHandler = (e) => {
    e.preventDefault();
    const email = document.getElementById('Remail');
    const name = document.getElementById('Rname');
    const password = document.getElementById('Rpassword');
    const repeatPassword = document.getElementById('RrepeatPassword');

    const emailValue = email.value;
    const nameValue = name.value;
    const passwordValue = password.value;
    const repeatPasswordValue = repeatPassword.value;

    if (
      emailValue.trim().length < 2 ||
      passwordValue.trim().length < 6 ||
      nameValue.trim().length < 2 ||
      repeatPasswordValue !== passwordValue
    )
      return alert('error in register form');

    dispatch(startRegister(emailValue, passwordValue, nameValue));

    email.value = '';
    name.value = '';
    password.value = '';
    repeatPassword.value = '';
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={onLoginSubmitHandler}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Email"
                name="email"
                id="lemail"
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name="password"
                id="lpassword"
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={onSubmitRegisterHandler}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Name"
                id="Rname"
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Email"
                id="Remail"
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                id="Rpassword"
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repeat password"
                name="repeatPassword"
                id="RrepeatPassword"
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Register" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
