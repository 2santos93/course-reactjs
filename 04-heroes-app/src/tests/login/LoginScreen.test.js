import { mount } from "enzyme";
import { AuthContext } from "../../auth/AuthContext";
import { LoginScreen } from "../../components/login/LoginScreen";

describe("LoginScreen test", () => {
  const state = {
    dispatch: jest.fn(),
  };

  const history = {
    replace: jest.fn()
  }

  const wrapper = mount(
    <AuthContext.Provider value={state}>
      <LoginScreen history={history} />
    </AuthContext.Provider>
  );

  test("should render ok", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should do login action", () => {
    wrapper.find('button').simulate('click');

    expect(state.dispatch).toHaveBeenCalled();
    expect(history.replace).toHaveBeenCalledWith('/');
  });
});
