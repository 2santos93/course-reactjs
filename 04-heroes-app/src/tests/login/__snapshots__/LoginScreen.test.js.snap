// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen test should render ok 1`] = `
Array [
  <h1>
    Login
  </h1>,
  <hr />,
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>,
]
`;
